[Creating Test] Creating IDM_9_1 python3 test module and consolidate IDM base-class inheritance

j-ororke opened the pull request:
#### Summary

- Creating python3 test module TC_IDM_9_1 for attribute and command data type constraint validation.
- Found and isolated 2 attributes that were not having their constraints enforced: 
  - ImageRotation of the Camera AV Stream Management cluster not respecting max constraint of 359 as able to write value 360
  - DefaultOpenDuration attribute of the Valve Configuration and Control cluster not respecting min constraint of 1 as able to write 0 
  - Flattens IDM base-class inheritance: IDMBaseTest now extends BasicCompositionTests, and TC_IDM_2_2, TC_IDM_3_2, TC_IDM_4_3, TC_IDM_5_2, TC_IDM_9_1 drop the redundant second base.

#### Related issues

- Follows test plan PR [5755](https://github.com/CHIP-Specifications/chip-test-plans/pull/5755)
- ~~Attribute constraint enforcement issue [42416](https://github.com/project-chip/connectedhomeip/issues/42416)~~(Fixed)

#### Testing

- Command used for testing in WSL: 
```
scripts/tests/run_python_test.py --factory-reset --app out/linux-x64-all-clusters/chip-all-clusters-app --app-args "--discriminator 1234 --KVS kvs1" --script-args "--storage-path admin_storage.json --commissioning-method on-network --discriminator 1234 --passcode 20202021 --PICS src/app/tests/suites/certification/ci-pics-values" --script src/python_testing/TC_IDM_9_1.py 
```

j-ororke committed:
[Creating Test] Creating IDM_9_1 python3 test module
- Follows test plan PR [5755](https://github.com/CHIP-Specifications/chip-test-plans/pull/5755)
- Command used for testing:
Found during testing 2 attributes that are not respecting constraints: ImageRotation of the Camera AV Stream Management cluster not respecting max constraint of 359 as able to write value 360, and DefaultOpenDuration attribute of the Valve Configuration and Control cluster not respecting min constraint of 1 as able to write 0.

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting error

j-ororke committed:
Updating logic to use command_guard for test steps 1a-1c and some minor cleanup

restyled-commits committed:
Restyled by ruff

j-ororke commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,506 @@
+                    optionsMask=0,
+                    optionsOverride=0
+                )
+                await self.default_controller.SendCommand(nodeId=self.dut_node_id, endpoint=1, payload=cmd)
```
Updated

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Fixing issue with ImageRotation and DefaultOpenDuration attributes constraints not being enforced:
- Fixed possible typo in SetImageRotation() in CameraAVStreamManagementCluster C++ module
- Added write handler to enforce min=1 constraint on DefaultOpenDuration in valve-configuration-and-control-cluster C++ module

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

j-ororke commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,506 @@
+                f"{cluster_name_no_spaces}.xml",
+                f"{cluster_name_with_dashes}-Cluster.xml",
+                f"{cluster_name_no_spaces}-Cluster.xml",
+            ]
```
These are not dupes, all 4 patterns are distinct and necessary to handle different XML filenames.

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Update TC_IDM_9_1.py

Updating imports due to framework changes

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Merge remote-tracking branch 'upstream/master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Restore DefaultOpenDuration min constraint in C++ ValveConfigurationAndControlCluster

restyled-commits committed:
Restyled by clang-format

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Moving functions out of IDM_9_1 so that they can be shared in spec_parsing and idm_support modules for future use

restyled-commits committed:
Restyled by ruff

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,499 @@
+
+
+class TC_IDM_9_1(BasicCompositionTests):
+    def _get_xml_path(self, cluster_id: int) -> Optional[str]:
```
Hi Cecille, 
I merged our changes to recent master in so that we could move the following functions into idm_support support module from IDM_9_1 test module:

- checkable_attributes()         
- resolve_dynamic_constraint()  
- generate_constraint_violation() 
- test_attribute_constraint()  

Then I moved the following functions into the spec_parsing module as you have requested above:

- get_xml_path()            
- parse_constraint_references()  
- parse_attribute_constraints() 

Then refactored IDM_9_1 test module to make the test code work with the now shared functions.

j-ororke committed:
Resolve type errors found in spec_parsing module

j-ororke committed:
Update idm_support.py

Adding back @staticmethod decorator as without it IDM_4_2 appears to fail in CI.

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/spec_parsing.py:
```diff
@@ -1687,3 +1687,169 @@ def dm_from_spec_version(specification_version: uint) -> PrebuiltDataModelDirect
+        min_elem = constraint_elem.find('./min')
+        if min_elem is not None:
+            if 'value' in min_elem.attrib:
+                constraints['min'] = int(min_elem.attrib['value'])
```
instead of using string tags in a separate list, please create classes or at least an enum for the contraints. 

The same constraint parsing can be used for attributes and also for device types eventually (@AryaHassanli FYI)

If you look in the parse_attributes file, there's functions for parsing the conformance and access. Matching that pattern here makes sense - when those are called, you already have direct access to the cluster element, so you can eliminate all the code above for finding and parsing the xml from file.

j-ororke committed:
Integrate constraint parsing into cluster loading
- Parse constraints once during cluster XML loading and store in XmlAttribute.constraints
- Handle float constraint values (e.g., 0.01) in addition to integers

j-ororke committed:
Refactor constraint parsing to use dataclasses and integrate into ClusterParser:
- Replace dict-based constraints with Constraints and ConstraintReference dataclasses
- Parse constraints directly in parse_attributes() method using element access
- Make constraint parsing reusable for device types via XmlDataTypeComponent
- Fix hex value parsing (e.g., '0x001F') in numeric constraints

restyled-commits committed:
Restyled by ruff

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving issue noticed with data model directory CI check

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Resolving linting errors

j-ororke committed:
Resolve mypy type error

j-ororke committed:
Resolve mypy type error round 2

j-ororke committed:
Update TestSpecParsingDatatype.py to using constraints dataclass instead of dictionary

j-ororke committed:
Update src/python_testing/TC_IDM_9_1.py

Co-authored-by: Raul Marquez <130402456+raul-marquez-csa@users.noreply.github.com>

raul-marquez-csa commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,222 @@
+            cmd = Clusters.GeneralCommissioning.Commands.SetRegulatoryConfig(
+                newRegulatoryConfig=Clusters.GeneralCommissioning.Enums.RegulatoryLocationTypeEnum.kIndoor,
+                countryCode="USA",  # Out of range (must be 2 chars)
+                breadcrumb=0
```
Has default value 0, can be ignored.

j-ororke committed:
Update TC_IDM_9_1.py

Updating to remove setting default values for args as not needed.

j-ororke committed:
Update TC_IDM_9_1.py

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

raul-marquez-csa reviewed (approved):
Comments  addressed.
LGTM.

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke commented on src/python_testing/matter_testing_infrastructure/matter/testing/spec_parsing.py:
```diff
@@ -774,6 +845,172 @@ def parse_features(self) -> dict[uint, XmlFeature]:
+            return None
+
+        # Helper to parse numeric values (handles both int and float)
+        def parse_numeric_value(value_str: str) -> Optional[Union[int, float]]:
```
Thanks, good catch. 
I have consolidated that into a single module-level helper, _parse_numeric_constraint_value, and updated both _parse_field_constraints and parse_attribute_constraints to use it.

I left the separate parse_int_value helper in _parse_field_constraints as-is, since that one is intentionally doing strict integer parsing for count values, which is a little different. This seemed like the cleanest way to remove duplication while keeping the count-specific handling intact.

j-ororke committed:
Refactor numeric constraint parsing in spec_parsing

Consolidate duplicate parse_value / parse_numeric_value helpers from _parse_field_constraints and parse_attribute_constraints into a single module-level _parse_numeric_constraint_value helper.

Pure refactor, no behavior change.

j-ororke commented on src/python_testing/support_modules/idm_support.py:
```diff
@@ -371,6 +374,139 @@ def verify_all_endpoints_clusters(self, read_response: dict):
+
+        return None
+
+    async def check_attribute_constraint(self, attr_info: dict, constraints: Constraints) -> bool:
```
Yes, that was something I originally created in the test. 
I went ahead and converted it into a WritableAttributeInfo dataclass in idm_support.py here now alongside the helpers that consume it, and updated generate_constraint_violation, check_attribute_constraint, and the collection loop in TC_IDM_9_1.py. 
It uses typed fields now instead of string-keyed lookups, which felt cleaner and less error prone.

j-ororke committed:
Replace attr_info dict with WritableAttributeInfo dataclass

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke commented on src/python_testing/matter_testing_infrastructure/matter/testing/spec_parsing.py:
```diff
@@ -774,6 +845,172 @@ def parse_features(self) -> dict[uint, XmlFeature]:
+        max_length_elem = constraint_elem.find('./maxLength')
+        if max_length_elem is not None and 'value' in max_length_elem.attrib:
+            with contextlib.suppress(ValueError):
+                # Value is not a simple integer, ignore for now
```
Good question. I checked all bundled data model XMLs from 1.0 through 1.6, including globals/Structs.xml, and every <minLength> / <maxLength> value I found is currently a plain integer literal.

So in practice, the suppress(ValueError) path appears to be defensive only right now. I added it mainly for symmetry with the min/max/minCount/maxCount parsing above, where non-integer expressions and references can appear. 

That said, I agree it should be tracked instead of silently ignored if this ever changes. I’ll open a follow-up issue to document the current state and decide whether we should support references for minLength/maxLength or make this a hard failure. I’ll link the issue here once filed.

j-ororke replied:
Follow-Up PR has been created here to help track this: https://github.com/project-chip/connectedhomeip/issues/71862

j-ororke commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,217 @@
+
+        # Step 1a: Test uint16 constraint validation using ColorControl clusters MoveToColorTemperature command
+        self.step("1a")
+        if await MatterBaseTest.command_guard(self, endpoint=self.endpoint, command=Clusters.ColorControl.Commands.MoveToColorTemperature):
```
Agreed, that makes sense. The hard-coded command steps are too specific to certain device types, so they can end up not exercising constraint validation on other DUTs.

Since removing steps 1a/1b/1c would also require a matching test plan update, I think the lower-risk path for this PR is to leave the current steps in place and add a TODO pointing to a follow-up issue.

I’ll open a follow-up issue for 1.6.1 to track the more systematic approach of selecting commands based on what the DUT actually exposes, then generating valid / invalid payloads from the parsed constraints.

Let me know if you’d prefer a different approach.

j-ororke replied:
Follow-up task has been created to keep track of this here: https://github.com/project-chip/connectedhomeip/issues/71864

cecille replied:
which test plan are you following? The one I have uses open commissioning window, which is mandatory for commissionable devices. It won't work for NTL, but that's rather their problem to deal with.

j-ororke replied:
Follow-up from Slack discussion:

I updated test steps 1b/1c to use Clusters.GeneralDiagnostics.Commands.TestEventTrigger(enableKey=...) for the octstr constraint case. Since EnableKey is mandatory and has a fixed 16-byte length, sending 15 or 17 bytes provides a deterministic way to exercise the constraint response. Thank you Cecille for the suggestion!

For test step 1a, still discussing the best command/field to use for a uint16 constraint case.


**Note:** Test plan PR **[5755](https://github.com/CHIP-Specifications/chip-test-plans/pull/5755)** updated to mention using a different command for test steps 1b and 1c.

ReneJosefsen replied:
I could not find this mentioned clearly in the spec, but what is the behavior if TestEventTriggersEnabled attribute is false? If it is the same as for the PayloadTestRequest command, then it would give a CONSTRAINT_ERROR even if the constraint is met? I have not checked the SDK.

While the command is mandated, it is not all device types which needs Test event triggers, so I suspect many devices coming to cert would have TestEventTriggersEnabled == false, so maybe this is not the best command to use for this case?

j-ororke replied:
@Renelosefsen good catch, thank you for bringing this up. I think you are right.

I took another look at the SDK code and HandleTestEventTrigger in GeneralDiagnosticsCluster.cpp routes wrong-length, all-zero-key, missing-delegate, and wrong-key cases to Status::ConstraintError through the same GetTriggerDelegateOnMatchingKey() helper [here.](https://github.com/project-chip/connectedhomeip/blob/d67b0db481617d861bb95def8d9b0e646eadd72e/src/app/clusters/general-diagnostics-server/GeneralDiagnosticsCluster.cpp#L119)

So on a DUT with TestEventTriggersEnabled == false, or where the TH does not have the matching 16-byte enable key, our 17-byte or 15-byte payload could get CONSTRAINT_ERROR for the wrong reason. The test would pass without actually validating length-constraint enforcement for the command field we care about.

That means TestEventTrigger.EnableKey does not look like a good choice for this cert test case after all. It has a similar overloaded error response issue as PayloadTestRequest. I’m going to look for a different octstr-with-length-constraint command/field that avoids this kind of overloaded response.

Thanks again for the review and the extra context here. Much better to catch this now.

j-ororke replied:
Small update after discussing this with @cecille during weekly standup.

The SDK behavior I mentioned above is still true, but Cecille clarified that TestEventTriggersEnabled is required to be enabled during cert events and testing. With that requirement in place, TestEventTrigger.EnableKey should still be valid for this test case.

So I think we can leave this as-is for now, unless there is another concern I am missing. 

Thanks again for calling this out.

j-ororke committed:
Added TODO for follow-up issue 71864 to test steps 1a, 1b, and 1c.

j-ororke committed:
Updating test steps 1b and 1c to using GeneralDiagnostics.TestEventTrigger, thanks to Cecille for the suggestion

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

j-ororke committed:
Removed test step 1a and reordered test steps

cecille commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,205 @@
+log = logging.getLogger(__name__)
+
+
+class TC_IDM_9_1(IDMBaseTest, BasicCompositionTests):
```
are thees both derived from matter base test? That seems odd. Do all the IDM tests require the wildcard? If so, would it make sense to just have the idm test inherit from basic composiition?

j-ororke replied:
Good call, I went ahead and did the consolidation here since the change was smallish. 
IDMBaseTest now extends BasicCompositionTests directly:
```
class IDMBaseTest(BasicCompositionTests):
    """Base test class for IDM tests with shared functionality."""
```
And the IDM tests that previously used the (IDMBaseTest, BasicCompositionTests) diamond have been simplified to just (IDMBaseTest):

- TC_IDM_2_2
- TC_IDM_3_2
- TC_IDM_4_3
- TC_IDM_5_2
- TC_IDM_9_1

TC_IDM_1_2, TC_IDM_1_4, and TC_IDM_4_2 were already on just IDMBaseTest and now transitively pick up BasicCompositionTests for free. Verified each of them still works with the wildcard-related symbols (setup_class_helper, endpoints_tlv, etc.) since those are now inherited via the chain. 
TC_IDM_2_3 is the only IDM test still on BasicCompositionTests directly; I left it as-is since it doesn't reference anything from IDMBaseTest, but happy to move it over if you'd prefer everything under one umbrella.

cecille commented on src/python_testing/TC_IDM_9_1.py:
```diff
@@ -0,0 +1,205 @@
+        # EnableKey is an octstr with allowed=16 (must be exactly 16 bytes); a 15-byte value violates the length constraint.
+        # TestEventTrigger lives on the root node (endpoint 0).
+        self.step("1b")
+        if await MatterBaseTest.command_guard(self, endpoint=0, command=Clusters.GeneralDiagnostics.Commands.TestEventTrigger):
```
I think when Rene was talking about test event triggers not being enabled, he was talking about them not being keyed. ie, assert that TestEventTriggersEnabled is true

j-ororke replied:
Thanks Cecille, that makes sense.

I think I was focused more on whether TestEventTriggersEnabled is expected to be enabled during cert testing, but I see the concern here is also that the test should explicitly guard/assert that precondition before using the command.

I updated the code to add an additional check for TestEventTriggersEnabled being true, similar to what is done in TC_TestEventTrigger.py [here](https://github.com/project-chip/connectedhomeip/pull/42411#discussion_r3268008555).

Thanks again for clarifying.

j-ororke committed:
Add additional check to test steps 1a and 1b to verify TestEventTriggersEnabled is True before running those test steps

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

j-ororke committed:
Merge branch 'master' into create_TC-IDM-9-1_python3-test-module

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

j-ororke committed:
IDMBaseTest in idm_support now extends BasicCompositionTests instead of MatterBaseTest, removing the diamond inheritance previously used by TC_IDM_2_2, TC_IDM_3_2, TC_IDM_5_2, and TC_IDM_9_1.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
